### DISCUSSION TITLE
Roadmap

### DISCUSSION THREAD

**Discussion by isixline** [archguard/archguard#5] [Show and tell]
## ArchGuard 1.0 Monolith application

this version is InnerSource only

## ArchGuard 2.0 Distributed (microservices) analysis

- [x] Multiple Language
   - [x] with Chapi
   - [x] API declaration analysis
       - [x] Java/Kotlin
       - [x] C#
   - [x] API call analysis
       - [x] Java
       - [x] Kotlin
       - [x] TypeScript  (50%)
- [x] Summary
    - [x] Language Summary
    - [x] Unstable file
    - [x] API list
- [x] HTTP API dependencies analysis
    - [x] Java/Kotlin
        - [x] resource: Spring
        - [x] demand: RESTTemplate
    - [x] JavaScript/Typescript
        - [x] demand: Axios
        - [x] demand: Umi-Request
- [x] Database analysis
   - [x] Java/Kotlin
      - [x] JDBI
      - [x] JPA
      - [x] Mybatis (XML)
- [x] Change analysis
    - [x] Git analysis
- [x] Debug log
    - [x] show log in frontend
- [x] Refactor
    - [x] Scanner
       - [x] Lint Rule Model
       - [x] Test Bad Smell
       - [x] Scanner Refactor to Plugins
- [x] Distributed linter
    - [x] API Linter（refs: [https://github.com/stoplightio/spectral](https://github.com/stoplightio/spectral)）
    - [x] Database Linter

## ArchGuard 3.0 Architecture Workbench

Goal：适应度函数驱动的架构治理

初步计算：

- 工作流引擎：参考 GitHub Action 的任务编排 + 插件化机制（registry）
- 更丰富的 feeder：在 ArchGuard 的 Analyser 基础上，提供更丰富的数据 API
- 灵活的数据查询：基于 MongoDB/MySQL 构建 DSL 数据查询接口
- 自定义适应度函数：基于 Kotlin REPL + 数学引擎，构建适应度函数
- 增强架构可视化功能：让团队能自定义指标可视化，构建适合于团队的架构数据大屏

Todos:

- [ ] Workflow Engine
   - [ ] ArchGuard Action
- [ ] More Feeder
    - [ ] Metric Feeder
    - [ ] Architecture Linter (refs: [https://github.com/modernizing/guarding](https://github.com/modernizing/guarding)
- [ ] Custom rule engine 
   - [ ] Rule DSL
   - [ ] Architecture DSL
   - [ ] Layered DSL
- [ ] DSL support
  - [ ] online DSL editor
- [ ] Fitness Engine

**Comment by wizardbyron** [comment]
Will archguard support monorepo ?

**Reply by phodal** [reply]
有限的支持。archguard 自己就是 monorepo

**Comment by huiwenhan** [comment]
any update about the roadmap ?

**Reply by phodal** [reply]
Gen AI has high priority, we working on other projects: https://github.com/unit-mesh/auto-dev
